provisioner/ansible/provisioner.go

```diff
@@ -307,50 +300,117 @@ func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.C
+	}
+	tf.Close()
+	p.config.InventoryFile = tf.Name()
+
+	return nil
+}
+
+func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.Communicator, generatedData map[string]interface{}) error {
+	ui.Say("Provisioning with Ansible...")
+	// Interpolate env vars to check for generated values like password and port
+	p.generatedData = generatedData
+	p.config.ctx.Data = generatedData
+	for i, envVar := range p.config.AnsibleEnvVars {
+		envVar, err := interpolate.Render(envVar, &p.config.ctx)
 		if err != nil {
-			return fmt.Errorf("Error preparing inventory file: %s", err)
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
-		defer os.Remove(tf.Name())
-
-		host := fmt.Sprintf("%s ansible_host=127.0.0.1 ansible_user=%s ansible_port=%d\n",
-			p.config.HostAlias, p.config.User, p.config.LocalPort)
-		if p.ansibleMajVersion < 2 {
-			host = fmt.Sprintf("%s ansible_ssh_host=127.0.0.1 ansible_ssh_user=%s ansible_ssh_port=%d\n",
-				p.config.HostAlias, p.config.User, p.config.LocalPort)
+		p.config.AnsibleEnvVars[i] = envVar
+	}
+	// Interpolate extra vars to check for generated values like password and port
+	for i, arg := range p.config.ExtraArguments {
+		arg, err := interpolate.Render(arg, &p.config.ctx)
+		if err != nil {
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
+		p.config.ExtraArguments[i] = arg
+	}
+
+	// Set up proxy if there's no host IP to access, regardless of user config.
+	hostIP := generatedData["Host"]
+	if hostIP == "" && p.config.UseProxy.False() {
+		ui.Error("Warning: use_proxy is false, but instance does" +
+			" not have an IP address to give to Ansible. Falling back" +
+			" to use localhost proxy.")
+		p.config.UseProxy = confighelper.TriTrue
+	}
 
-		w := bufio.NewWriter(tf)
-		w.WriteString(host)
-		for _, group := range p.config.Groups {
-			fmt.Fprintf(w, "[%s]\n%s", group, host)
+	privKeyFile := ""
+	if !p.config.UseProxy.False() {
+		// We set up the proxy if useProxy is either true or unset.
+		err, privKeyFile := p.setupAdapterFunc(ui, comm)
+		if err != nil {
+			return err
 		}
 
-		for _, group := range p.config.EmptyGroups {
-			fmt.Fprintf(w, "[%s]\n", group)
+		defer func() {
+			log.Print("shutting down the SSH proxy")
+			close(p.done)
+			p.adapter.Shutdown()
+		}()
+
+		go p.adapter.Serve()
+
+		// Remove the private key file
+		if len(privKeyFile) > 0 {
+			defer os.Remove(privKeyFile)
 		}
+	} else {
+		ui.Message("Not using Proxy adapter for Ansible run...")
+	}
 
-		if err := w.Flush(); err != nil {
-			tf.Close()
-			return fmt.Errorf("Error preparing inventory file: %s", err)
+	if len(p.config.InventoryFile) == 0 {
+		if !p.config.UseProxy.False() {
+			err := p.createInventoryFile("127.0.0.1", p.config.LocalPort)
+			if err != nil {
+				return err
+			}
+		} else {
+			err := p.createInventoryFile(generatedData["Host"].(string),
+				int(generatedData["Port"].(int64)))
+			if err != nil {
+				return err
+			}
```

Comment:
NIT: I would have made of this block something simpler :
```go
	hostIp, hostPort := "127.0.0.1", p.config.LocalPort
	if p.config.UseProxy.False() == true {
		hostIp, hostPort = generatedData["Host"].(string), int(generatedData["Port"].(int64)))
	}
	err := p.createInventoryFile(hostIp, hostPort)
```

Reply:
Yep, this is better.